Add some docs on the Digital Identity integration, and debugging it (#305)
alphagov/account-api: docs/digital-identity.md

# Digital Identity

This document gets into some of the details of the integration with
Digital Identity, and where it can go wrong.  For a high-level
overview, see [the developer docs][].

[the developer docs]: https://docs.publishing.service.gov.uk/manual/govuk-account.html

## Big picture

> The `OidcUser` model, which represents a user who has authenticated

Reviewer: ```suggestion
The [`OidcUser` model](https://github.com/alphagov/account-api/blob/main/app/models/oidc_user.rb), which represents a user who has authenticated
```

via Digital Identity, holds four special pieces of data:

- The `sub`, or "subject identifier", is a unique identifier assigned
  by Digital Identity when the account is created, and never changes.
  Other data, like an email address, can change: so this is what we
  use to identify users.

- The `legacy_sub` is the user's subject identifier from the old
  [account manager prototype][].  Only accounts which were created
  before the Digital Identity migration have this set.  We use this to
  join up old data.

- The `email` is the user's current email address.  It is set by
  Digital Identity calling the `/api/oidc-users/:subject_identifier`
  endpoint.

> - The `email_verified` is whether the user's current email address is

Reviewer: ```suggestion
- The `email_verified` is a boolean denoting whether the user's email address is
```

  verified.  It is set by Digital Identity calling the
  `/api/oidc-users/:subject_identifier` endpoint.  As Digital Identity
  require immediate verification of email addresses, this is always
  true.

Communication flows in both directions between account-api and Digital
Identity:

- When a user is updated or deleted, Digital Identity call the

>   endpoints in the `Internal::OidcUsersController`.

Reviewer: ```suggestion
  endpoints in [`Internal::OidcUsersController`](https://github.com/alphagov/account-api/blob/main/app/controllers/internal/oidc_users_controller.rb).
```

- When a user logs in, we call Digital Identity from the

>   `Internal::AuthenticationController`.

Reviewer: ```suggestion
  [`Internal::AuthenticationController`](https://github.com/alphagov/account-api/blob/main/app/controllers/internal/authentication_controller.rb).
```

  subdomain of that), then it's Digital Identity.
- If account-api is persistently unable to connect to
  `https://oidc.account.gov.uk`, then it's Digital Identity.

If the problem is on the Digital Identity side, or if you need support
from them, ask for help in `#di-authentication` on Slack.

If the problem is not clearly to do with Digital Identity, read on.

### Email address out of date

Find out from the user what the email address should be and what it's
currently showing as.

You can find their user record with:

```ruby
user = OidcUser.find_by(email: "currently showing address")
```

You can also see if there is a user with the email address they are
trying to change to:

```ruby
other_user = OidcUser.find_by(email: "desired new email address")
```

Digital Identity enforces that email addresses are unique, so we
shouldn't ever have two users with the same email address.

If there is another user with the desired email address, something may
have gone wrong with updating *their* user record, which will have to
be resolved before this user can be updated.

#### Check Sentry for `CapturedSensitiveException` events

> 

Reviewer: I can't find Account API in Sentry, presumably because I don't have access. Should access be granted to the [`#govuk-platform-health` team](https://sentry.io/settings/govuk/teams/govuk-platform-health-team/projects/) like our other GOV.UK apps?

Also could you link to the Sentry project from this section?

failed due to some error which wasn't reported.

Check with them if the API call was made and if it was successful.  If
the call was made but failed, then continue debugging account-api.

### Can't log in

Find out from the user what the email address they've used for their
account is.

You can find their user record with:

```ruby
user = OidcUser.find_by(email: "address")
```

#### Can't log in for the first time

If there is no `OidcUser` record for that email address, this is the
user's first time logging in.  When a user first logs in, we create
their `OidcUser` record and fetch their email address.

##### Check Sentry for `ActiveRecord::RecordNotUnique` events

An exception is raised if two `OidcUsers` have the same subject
identifier:

```
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_oidc_users_on_sub"`
DETAIL:  Key (sub)=(foo) already exists.
```

This should never happen in production.

As subject identifiers come from Digital Identity, contact them for
support and give the duplicate subject identifier (`foo` in the
example above).

##### Check Sentry for `CapturedSensitiveException` events

> Try the debugging steps described above under "email address out of

Reviewer: ```suggestion
Try the [debugging steps](#check-sentry-for-capturedsensitiveexception-events) described earlier (under "email address out of
```

date": if there is another user record with the user's email address,
creating their account will fail.

##### Check Sentry for `UserDestroyed` events

An exception is raised if a user logs into an account which Digital
Identity has previously told us is deleted:

```
AccountSession::UserDestroyed (AccountSession::UserDestroyed):

app/lib/account_session.rb:27:in `initialize'
app/controllers/internal/authentication_controller.rb:22:in `new'
app/controllers/internal/authentication_controller.rb:22:in `callback'
```

This is because we create a `Tombstone` record when a user is
destroyed.  This record holds the subject identifier, and we use it to
end any active sessions that user still has.

If a user logs into an account that has a subject identifier with a
corresponding `Tombstone` record, Digital Identity have re-used a
subject identifier, or un-deleted the account.  We currently assume
this does not happen.

Removing the `Tombstone` record will allow the user to log in.  But if
this is an account which has been un-deleted, any data we held on them
will be lost.

#### Can't log in at other times

> Check that Digital Identity is up and running, and ask if they are

Reviewer: How?

seeing an elevated rate of invalid authorization codes or access
tokens from GOV.UK.

If Digital Identity is experiencing a problem, we just have to wait
for them to fix it.

Otherwise, we have a bug in account-api, or perhaps elsewhere in the
GOV.UK stack.  One thing to try would be to clear our cached copy of
the Digital Identity OIDC configuration:

```ruby
Rails.cache.clear
```

> Otherwise, there isn't any more specific guidance I can give here.

Reviewer: ```suggestion
```

Try to avoid use of first person in docs. Also not much value to this closing sentence.

##### Check Sentry for `CapturedSensitiveException` events

> Try the [debugging steps](#check-sentry-for-capturedsensitiveexception-events)

Reviewer: ~~Hmmm this anchor link references the section we're already in?~~

Oh wait, we have two identical headings! In which case, I'd be tempted to put a `<a name="debugging-steps"></a>` in the appropriate place and then link to `#debugging-steps` here.

Author: Thanks, I've fixed this and also made some small tweaks.
